File: julia/Euler/src/bin/p0076.jl

# project euler: problem 76

#=
  another version of problem 31

  coins: 1, 2, 3, ..., 99
  total: 100
=#

module Prob0076

function solve_0076(coins::Vector{Int} = collect(1:99), target::Int = 100)
    tbl = zeros(Int, target + 1)
    tbl[1] = 1

    for c in coins
        for i in c:target
            tbl[i + 1] += tbl[i - c + 1]
        end
    end
    tbl[end]
end

end #module

using .Prob0076: solve_0076
export solve_0076

File: clojure/src/project_euler/sol/p0076.clj
(ns project-euler.sol.p0076)

;;;; Another version of problem 31
;;;;
;;;;   coins: 1, 2, 3, ..., 99
;;;;   total: 100

(defn solve
  ([]
   (solve (range 1 100) 100))
  ([coins target]
   (let [tbl (long-array (inc target) 0)]
     (aset tbl 0 1)
     (doseq [c coins
             i (range c (inc target))]
       (aset tbl i (+ (aget tbl i) (aget tbl (- i c)))))
     (aget tbl target))))
